src/cascadia/TerminalControl/TermControl.cpp

```diff
@@ -2526,16 +2527,55 @@ namespace winrt::Microsoft::Terminal::Control::implementation
 
             if (items.Size() > 0)
             {
-                std::wstring allPaths;
-                for (auto item : items)
+                std::vector<std::wstring> fullPaths;
+
+                // GH#14628: Workaround for GetStorageItemsAsync() only returning 16 items
+                // at most when dragging and dropping from archives (zip, 7z, rar, etc.)
+                if (items.Size() == 16 && e.DataView().Contains(winrt::hstring{ L"FileDrop" }))
                 {
-                    // Join the paths with spaces
-                    if (!allPaths.empty())
+                    auto fileDropData = co_await e.DataView().GetDataAsync(winrt::hstring{ L"FileDrop" });
+                    if (fileDropData != nullptr)
+                    {
+                        auto stream = fileDropData.as<IRandomAccessStream>();
+                        stream.Seek(0);
+
+                        const uint32_t streamSize = gsl::narrow_cast<uint32_t>stream.Size();
+                        const Buffer buf(streamSize);
+                        const auto buffer = co_await stream.ReadAsync(buf, streamSize, InputStreamOptions::None);
+
+                        const HGLOBAL hGlobal = buffer.data();
+                        const auto count = DragQueryFile((HDROP)hGlobal, 0xFFFFFFFF, nullptr, 0);
+                        fullPaths.reserve(count);
+
+                        for (unsigned int i = 0; i < count; i++)
+                        {
+                            WCHAR szPath[MAX_PATH];
+                            const auto charsCopied = DragQueryFile((HDROP)hGlobal, i, szPath, MAX_PATH);
+
+                            if (charsCopied > 0)
+                            {
+                                fullPaths.push_back(std::wstring{ szPath });
+                            }
+                        }
+                    }
+                }
+                else
+                {
+                    fullPaths.reserve(items.Size());
+                    for (auto item : items)
                     {
-                        allPaths += L" ";
+                        fullPaths.push_back(std::wstring{ item.Path() });
                     }
+                }
 
-                    std::wstring fullPath{ item.Path() };
+                std::wstring allPathsString;
+                for (auto fullPath : fullPaths)

```

Comment:
```suggestion
                for (auto& fullPath : fullPaths)
```

No point taking an extra copy if we don't need to!